smart-contract-verifier/src/solidity/bytecode.rs

```diff
@@ -0,0 +1,618 @@
+                metadata_length_raw,
+                ..
+            } => metadata_raw.len() + metadata_length_raw.len(),
+        }
+    }
+}
+
+/// Encapsulates result of local source code compilation.
+/// Splits compiled creation transaction input.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct LocalBytecode {
+    bytecode: Bytecode,
+    creation_tx_input_parts: Vec<BytecodePart>,
+}
+
+impl LocalBytecode {
+    /// Initializes a new [`LocalBytecode`] struct.
+    ///
+    /// `bytecode` and `bytecode_modified` must differ only in metadata hashes.
+    ///
+    /// Any error here is [`VerificationErrorKind::InternalError`], as both original
+    /// and modified bytecodes are obtained as a result of local compilation.
+    pub fn new(
+        bytecode: Bytecode,
+        bytecode_modified: Bytecode,
+    ) -> Result<Self, VerificationErrorKind> {
+        let creation_tx_input_parts = Self::split(
+            &bytecode.creation_tx_input,
+            &bytecode_modified.creation_tx_input,
+        )?;
+
+        Ok(Self {
+            bytecode,
+            creation_tx_input_parts,
+        })
+    }
+
+    pub fn creation_tx_input(&self) -> &Bytes {
+        &self.bytecode.creation_tx_input
+    }
+
+    pub fn creation_tx_input_parts(&self) -> &Vec<BytecodePart> {
+        &self.creation_tx_input_parts
+    }
+
+    /// Splits bytecode onto [`BytecodePart`]s using bytecode with modified metadata hashes.
+    ///
+    /// Any error here is [`VerificationErrorKind::InternalError`], as both original
+    /// and modified bytecodes are obtained as a result of local compilation.
+    fn split(
+        raw: &Bytes,
+        raw_modified: &Bytes,
+    ) -> Result<Vec<BytecodePart>, VerificationErrorKind> {
+        if raw.len() != raw_modified.len() {
+            return Err(VerificationErrorKind::InternalError(format!(
+                "bytecode and modified bytecode length mismatch: {}",
+                Mismatch::new(raw.len(), raw_modified.len())
+            )));
+        }
+
+        let parts_total_size = |parts: &Vec<BytecodePart>| -> usize {
+            parts.iter().fold(0, |size, el| size + el.size())
+        };
+
+        let mut result = Vec::new();
+
+        let mut i = 0usize;
+        while i < raw.len() {
+            let decoded = Self::parse_bytecode_parts(&raw.slice(i..raw.len()), &raw_modified[i..])?;
+            let decoded_size = parts_total_size(&decoded);
+            result.extend(decoded);
+
+            i += decoded_size;
+        }
+
+        Ok(result)
+    }
+
+    /// Finds the next [`BytecodePart`]s into a series of bytes.
+    ///
+    /// Parses at most one [`BytecodePart::Main`] and one [`BytecodePart::Metadata`].
+    fn parse_bytecode_parts(
+        raw: &Bytes,
+        raw_modified: &[u8],
+    ) -> Result<Vec<BytecodePart>, VerificationErrorKind> {
+        let mut parts = Vec::new();
+
+        let len = raw.len();
+
+        let mut i = 0usize;
+        while i < len {
```

Comment:
let's remake this into more readable thing
basically you're searching for the common prefix and then do stuff with that. but you're doing stuff "while searching"
imo it will be much better if we decompose this stuff
1. Find common prefix, get the index of the last byte
2. Then do stuff
and we probably can use find methods on iterators (didn't test that but something along those lines)
```rust
let index = raw.iter().zip(raw_modified.iter()).enumerate().find(|(index, (left, right))| left != right).0;
// or at least one single for/while loop with only searching
```